spiffe/id.go

```diff
@@ -0,0 +1,242 @@
+func ValidateID(spiffeID string, mode ValidationMode) error {
+	_, err := ParseID(spiffeID, mode)
+	return err
+}
+
+// ValidateIDURL validates the SPIFFE ID according to the SPIFFE
+// specification, namely:
+// - spiffe id is not empty
+// - spiffe id is a valid url
+// - scheme is 'spiffe'
+// - user info is not allowed
+// - host is not empty
+// - port is not allowed
+// - query values are not allowed
+// - fragment is not allowed
+// - path does not start with '/spire' since it is reserved for agent, server, etc.
+// In addition, the validation mode is used to control what kind of SPIFFE ID
+// is expected.
+// For more information:
+// [https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md]
+func ValidateIDURL(id *url.URL, mode ValidationMode) error {
+	options := mode.validationOptions()
+
+	validationError := func(format string, args ...interface{}) error {
+		var kind string
+		switch options.idType {
+		case trustDomainId:
+			kind = "trust domain "
+		case workloadId:
+			kind = "workload "
+		}
+		return fmt.Errorf("invalid %sSPIFFE ID %q: "+format,
+			append([]interface{}{kind, id.String()}, args...)...)
+	}
+
+	if id == nil || *id == (url.URL{}) {
+		return validationError("SPIFFE ID is empty")
+	}
+
+	// General validation
+	switch {
+	case strings.ToLower(id.Scheme) != "spiffe":
+		return validationError("invalid scheme")
+	case id.User != nil:
+		return validationError("user info is not allowed")
+	case id.Host == "":
+		return validationError("trust domain is empty")
+	case id.Port() != "":
+		return validationError("port is not allowed")
+	case id.Fragment != "":
+		return validationError("fragment is not allowed")
+	case id.RawQuery != "":
+		return validationError("query is not allowed")
+	}
+
+	// trust domain validation
+	if options.trustDomainRequired {
+		if options.trustDomain == "" {
+			return errors.New("trust domain to validate against cannot be empty")
+		}
+		if id.Host != options.trustDomain {
+			return fmt.Errorf("%q does not belong to trust domain %q", id, options.trustDomain)
+		}
+	}
+
+	// id type validation
+	switch options.idType {
+	case anyId:
+	case trustDomainId:
+		if id.Path != "" {
+			return validationError("path is not empty")
+		}
+	case workloadId:
+		if id.Path == "" {
+			return validationError("path is empty")
+		}
+	default:
+		return validationError("internal error: unhandled id type %v", options.idType)
+	}
+
+	return nil
+}
+
+// ParseID parses the SPIFFE ID and makes sure it is valid according to
+// the specified validation mode.
+func ParseID(spiffeID string, mode ValidationMode) (*url.URL, error) {
+	u, err := url.Parse(spiffeID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid SPIFFE ID: %v", err)
+	}
+
+	if err := ValidateIDURL(u, mode); err != nil {
+		return nil, err
+	}
+
+	return normalizeURL(u), nil
+}
+
+// ValidationMode is used to validate the SPIFFE ID being parsed/validated
+// should have.
+type ValidationMode interface {
```

Comment:
I think the comment is missing something - that or I simply don't understand it :)

Does this need to be exported? Doesn't look like anything related to it is

Reply:
whoops!

Reply:
It doesn't need to be exported but looks funny to me that way since it shows up in exported function signatures